### PR TITLE
add dispatch hooks for internal integrator fields

### PR DESCRIPTION
**Pull request by isaacsas**
This uses dispatch hooks for integrator tstop queries to enable other time steppers than ODE/SDEs (i.e. JumpProcesses) to work for periodic callbacks too.

Before merging needs the following merged with package releases
- https://github.com/SciML/DiffEqBase.jl/pull/1071 
- https://github.com/SciML/OrdinaryDiffEq.jl/pull/2377 
- https://github.com/SciML/StochasticDiffEq.jl/pull/580
- https://github.com/SciML/Sundials.jl/pull/459

### CODE REVIEW

File: src/iterative_and_periodic.jl [hunk 1]
```diff
@@ -177,8 +182,7 @@ end
     # Checking for floating point equality is OK here as `DifferentialEquations.jl`
     # sets the time exactly to the final time in the last iteration
     return integrator.t == last(integrator.sol.prob.tspan) ||
-           isempty(integrator.opts.tstops) ||
```
Comment (isaacsas) [inline src/iterative_and_periodic.jl:180]: @ChrisRackauckas is this really needed? As far as I can tell `integrator.opts.tstops` is never empty for ODEs as it always contains `tspan[2]` too, whereas in SSAStepper we do not insert `tspan[2]` into it (we actually don't copy the user `tstops` over, which we probably should do in cases that we insert into it but that is another issue).

If I remove this check everything seems to work ok in my limited testing for both jumps and ODEs (with the JumpProcesses PR I'll open).

Comment (isaacsas) [reply src/iterative_and_periodic.jl:180]: i.e. https://github.com/SciML/OrdinaryDiffEq.jl/blob/dc0e1e733fdd79a6c643cbd0c4f4be03c934cb91/lib/OrdinaryDiffEqCore/src/solve.jl#L609 adds `tspan[2]` into the internal ODE `tstops` which we never do for SSAStepper.

Comment (ChrisRackauckas) [reply src/iterative_and_periodic.jl:180]: There is a case where tstops is empty which is when a user manually takes control over stepping in the integrator interface. You can then have an `Inf` or `nothing` tspan. This is used for example in deployments of control systems, where the integrator is effectively set to just run forever based on a fixed timer for the control loop, reading in data and sending out a model-predicted reaction. This is somewhat different from the normal open source simulation usage but these deployments are something that's becoming a lot more common.

That said, this condition would then force the callback to end in that scenario, which seems like it's counter-productive to wanting that to never finish. So from what I can see, I think it's safe to remove this condition.

**Comment by ChrisRackauckas** [comment]
I think this is a good idea, it definitely abstracts away a few things that were originally more hacky into things that we can and should maintain.

File: src/iterative_and_periodic.jl [hunk 2]
```diff
@@ -35,20 +35,21 @@ function IterativeCallback(time_choice, user_affect!, tType = Float64;
...
         Okay yeah, this is nasty
         the comparer is always less than for type stability, so in order
         for this to actually check the correct direction we multiply by
         tdir
         =#
         tdir_tnew = integrator.tdir * tnew
+        tstops_array = get_tstops_array(integrator)
         for i in length(tstops):-1:1 # reverse iterate to encounter large elements earlier
-            if tdir_tnew < tstops.valtree[i] # TODO: relying on implementation details
+            if tdir_tnew < tstops_array[i] # TODO: relying on implementation details
```
Comment (ChrisRackauckas) [inline src/iterative_and_periodic.jl:48]:
```suggestion
            if tdir_tnew < tstops_array[i]
```

**Comment by isaacsas** [comment]
@ChrisRackauckas this seems like a depwarn related failure? https://github.com/SciML/DiffEqCallbacks.jl/actions/runs/10546737927/job/29218600378?pr=229#step:6:743

**Comment by isaacsas** [comment]
@ChrisRackauckas is there anyway to temporarily disable depwarn failures so we can get this merged? As far as I can tell, at this point this is only being held up due to them (I think tests passed locally last time I checked). 

I'd like to get this for JumpProcesses to finally fix the current issues with using PeriodicCallbacks there.

**Comment by isaacsas** [comment]
Not sure why the package resolver in the doc build is claiming OrdinaryDiffEq 6.88 doesn’t exist.

**Comment by isaacsas** [comment]
FWIW I can't reproduce this doc package resolver issue locally.

**Comment by isaacsas** [comment]
This looks good to go. As far as I can tell the doc failure is unrelated to this PR and seems resolver / caching related.

**Comment by ChrisRackauckas** [comment]
Yeah just a server issue, not going to worry about that right now.

**Pull request merged**